app/models/cost_query/cache.rb

```diff
@@ -0,0 +1,75 @@
+#-- copyright
+# OpenProject Reporting Plugin
+#
+# Copyright (C) 2010 - 2014 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#++
+
+module CostQuery::Cache
+  class << self
+
+    def check
+      reset! if reset_required?
+    end
+
+    def reset!
+      CostQuery::Filter.reset!
+      CostQuery::Filter::CustomFieldEntries.reset!
+      CostQuery::GroupBy.reset!
+      CostQuery::GroupBy::CustomFieldEntries.reset!
+
+      update_reset_on
+    end
+
+    protected
+
+    attr_accessor :custom_fields_updated_on,
+                  :custom_fields_id_sum
+
+    def invalid?
+      updated_on = fetch_custom_field_updated_at
+      id_sum = fetch_custom_fields_changed
+
+      custom_fields_updated_on != updated_on ||
+        custom_fields_id_sum != id_sum
+    end
+
+    def update_reset_on
+      return if caching_disabled?
+
+      self.custom_fields_updated_on = fetch_custom_field_updated_at
+      self.custom_fields_id_sum = fetch_custom_fields_changed
+    end
+
+    def fetch_custom_field_updated_at
+      WorkPackageCustomField.maximum(:updated_at)
+    end
+
+    def fetch_custom_fields_changed
+      WorkPackageCustomField.sum(:id) + WorkPackageCustomField.count
+    end
+
+    def caching_disabled?
+      !OpenProject::Configuration.cost_reporting_cache_filter_classes
+    end
+
+    def reset_required?
+      caching_disabled? || invalid?
```

Comment:
Wait a second...

When caching is **disabled** we will **always** `reset!` the cache? This sounds counter intuitive.

Your claim was disabling caching would remove the additional performance impact from going through a cache:
> and caching in the way it is implemented right now will thus only slow down the response.

As I understand it now (But I've not read all the code), you always reset the cache when it is disabled. I see how this resolves problems with wrong contents, because the cache is always fresh. I don't see how this avoids the impact of the cache in the first place.